agent/grpc-middleware/recovery.go

```diff
@@ -2,35 +2,22 @@ package middleware
 
 import (
 	recovery "github.com/grpc-ecosystem/go-grpc-middleware/recovery"
-	"github.com/hashicorp/go-hclog"
+	rate "github.com/hashicorp/consul/agent/consul/rate"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
 // PanicHandlerMiddlewareOpts returns the []recovery.Option containing
 // recovery handler function.
-func PanicHandlerMiddlewareOpts(logger Logger) []recovery.Option {
+func PanicHandlerMiddlewareOpts(logger rate.Logger) []recovery.Option {
 	return []recovery.Option{
-		recovery.WithRecoveryHandler(NewPanicHandler(logger)),
+		recovery.WithRecoveryHandler(recoveryHandlerFunc(logger)),
 	}
 }
 
-// NewPanicHandler returns a recovery.RecoveryHandlerFunc closure function
-// to handle panic in GRPC server's handlers.
-func NewPanicHandler(logger Logger) recovery.RecoveryHandlerFunc {
+func recoveryHandlerFunc(logger rate.Logger) recovery.RecoveryHandlerFunc {
 	return func(p interface{}) (err error) {
-		// Log the panic and the stack trace of the Goroutine that caused the panic.
-		stacktrace := hclog.Stacktrace()
-		logger.Error("panic serving grpc request",
-			"panic", p,
-			"stack", stacktrace,
-		)
-
-		return status.Errorf(codes.Internal, "grpc: panic serving request")
+		panicErr := rate.NewPanicHandler(logger)
```

Comment:
It feels a little odd to have the panic handler live in the `rate` package 🤔 

Personally, as it's just a couple of lines of code, I'd follow the [Go proverb](https://go-proverbs.github.io/) and just duplicate it:

> A little copying is better than a little dependency.

That said, if we do want to share it, maybe we could put it somewhere more generic like `agent/panic`?

Reply:
Hmm... `agent/panic` might be too broad of a scope for the limited use case it currently serves 🤔  WDYT?
I can certainly move it out if you feel strongly about it 🙂 

From my perspective, the rate package made sense because this is where `Allow` is implemented and the panic handler is limited to the rate-limiting atm.

Reply:
Ahh yes...! Thank you for keeping us honest @boxofrad 🙏 
I synced with @dhiaayachi and we decided to keep these separate.